### ISSUE TITLE
Some small ways to improve the code

### ISSUE THREAD

**Issue by Boscop** [dheijl/swyh-rs#22]
Thanks for helping me with the cpal wasapi loopback capture :)

I took a look at the code of swyh and noticed some small things:

1. Calling `get_output_audio_devices()` twice unnecessarily:
https://github.com/dheijl/swyh-rs/blob/d51bed261c104a5b3960162d86459436f3cfbacd/src/main.rs#L280
https://github.com/dheijl/swyh-rs/blob/d51bed261c104a5b3960162d86459436f3cfbacd/src/main.rs#L304

`cpal::Device` isn't `Clone` but you could reuse the result if you move
https://github.com/dheijl/swyh-rs/blob/d51bed261c104a5b3960162d86459436f3cfbacd/src/main.rs#L305-L307
to before
https://github.com/dheijl/swyh-rs/blob/d51bed261c104a5b3960162d86459436f3cfbacd/src/main.rs#L281-L287

2. You could eliminate `unsafe` [here](https://github.com/dheijl/swyh-rs/blob/d51bed261c104a5b3960162d86459436f3cfbacd/src/main.rs#L843-L849) by using [`once_cell`](https://docs.rs/once_cell/1.5.2/once_cell/index.html#lazy-initialized-global-data):
```rust
    static ONETIME_SW: OnceCell<()> = OnceCell::new();
    ONETIME_SW.get_or_init(|| {
        log("The wave_reader is receiving samples".to_string());
    })
```
(Btw, you could remove the dependency on `lazy_static` completely in favor of using `once_cell`.)

3. You could do [`let chunk = samples.to_vec();`](https://doc.rust-lang.org/nightly/std/primitive.slice.html#method.to_vec) here:
https://github.com/dheijl/swyh-rs/blob/d51bed261c104a5b3960162d86459436f3cfbacd/src/utils/rwstream.rs#L54-L56

4. You could do [`self.fifo.extend(chunk);`](https://doc.rust-lang.org/nightly/std/collections/struct.VecDeque.html#method.extend) here:
https://github.com/dheijl/swyh-rs/blob/d51bed261c104a5b3960162d86459436f3cfbacd/src/utils/rwstream.rs#L73-L75
(Also you can remove `continue;`.)
Btw, I haven't looked at your whole code, but why are you using `VecDeque` instead of `Vec`? 
It's usually slower due to cache-nonlocality. `Vec::push` is asymptotically `O(1)` (constant time) and has better cache locality & less overhead when iterating. Usually `Vec` is used for audio buffers.

5. You could theoretically use [`chunks_exact_mut`](https://doc.rust-lang.org/nightly/std/primitive.slice.html#method.chunks_exact_mut) here:
https://github.com/dheijl/swyh-rs/blob/d51bed261c104a5b3960162d86459436f3cfbacd/src/utils/rwstream.rs#L64

6. You could reuse this `Vec` between frames, to reduce allocations:
https://github.com/dheijl/swyh-rs/blob/d51bed261c104a5b3960162d86459436f3cfbacd/src/main.rs#L851

7. You could use `RwLock<HashMap<..>>` for [`CLIENTS`](https://github.com/dheijl/swyh-rs/blob/d51bed261c104a5b3960162d86459436f3cfbacd/src/main.rs#L98) instead of `Mutex<HashMap<..>>`, to allow multiple readers at the same time:
Either std: https://doc.rust-lang.org/nightly/std/sync/struct.RwLock.html
Or parking_lot: https://docs.rs/parking_lot/0.11.0/parking_lot/type.RwLock.html#differences-from-the-standard-library-rwlock
Or you could use a [concurrent hashmap](https://crates.io/search?q=concurrent%20hashmap), but I'm not sure what's best here since I haven't read the whole code..
Note that [`RwLock` is not always better](https://fy.blackhats.net.au/blog/html/2018/10/19/rust_rwlock_and_mutex_performance_oddities.html), at least std's `RwLock` on linux.

8. You can eliminate all the `unsafe` blocks around access to `CONFIG_CHANGED` like [this one](https://github.com/dheijl/swyh-rs/blob/d51bed261c104a5b3960162d86459436f3cfbacd/src/main.rs#L438-L458) or this one
https://github.com/dheijl/swyh-rs/blob/d51bed261c104a5b3960162d86459436f3cfbacd/src/main.rs#L224-L226
by using [`Rc<Cell<bool>>`](https://doc.rust-lang.org/book/ch15-05-interior-mutability.html) (or [`AtomicBool`](https://doc.rust-lang.org/nightly/std/sync/atomic/struct.AtomicBool.html) but the closures passed to [`handle2`](https://docs.rs/fltk/0.10.10/fltk/prelude/trait.WidgetBase.html#tymethod.handle2) don't need to be `Send`).

**Comment by dheijl** [comment]
Thanks a lot for the tips, I've just started using Rust and this is my first Rust program, this is very useful!

**Comment by dheijl** [comment]
1. Calling get_output_audio_devices() twice unnecessarily: this was on my low-priority to-do list as well, but thanks for reminding me!
2. You could eliminate unsafe here by using once_cell: good tip, I didn't know about once_cell.
3. You could do let chunk = samples.to_vec(); here: another good tip
4. You could do self.fifo.extend(chunk); here: again a good tip
5. You could theoretically use chunks_exact_mut here: true, but I like explicit code too
6. You could reuse this Vec between frames, to reduce allocations:: the Vec size is not fixed between invocations, both on Linux and Windows, or is there something I do not see?
 7. You could use RwLock<HashMap<..>> for CLIENTS instead of Mutex<HashMap<..>>: true, I prefer Mutex anyway as it is more than efficient enough in my experience, and very predictable in performance.
8. You can eliminate all the unsafe blocks around access to CONFIG_CHANGED: true, done, but it's jumping somewhat unnecessary through hoops to satisfy the Rust compiler in my opinion...

Thanks for taking the trouble to review my code, really appreciated.

**Comment by Boscop** [comment]
5. Re "chunks_exact_mut": Yeah, in this case, the code is easy enough to follow. I just meant to say that with `chunks_exact_mut` it would involve less mutable state, (no `i += 2;` necessary). I usually prefer the "more functional" version, using iterators instead of imperative code. They also get optimized/auto-vectorized better because the compiler can reason better about code that doesn't mutate state.

6. "the Vec size is not fixed between invocations, both on Linux and Windows, or is there something I do not see?":
You could still reuse it by calling `vec.clear(); vec.extend(samples.iter().map(|x| x.to_i16()));`.
So it will only allocate in the beginning (and after the len grows by more than factor 2, because it always keeps some extra capacity when reallocating to achieve amortized constant time for `push`).
So if the len only fluctuates a little, it won't reallocate because it has enough capacity.
Usually they say "you shouldn't block in the audio thread" and memory allocation can block:
http://www.rossbencina.com/code/real-time-audio-programming-101-time-waits-for-nothing
But it's usually about avoiding audio dropouts for output, whereas in this case it's an input stream. Maybe you haven't experienced dropouts because your system allocator is fast, but what if someone runs it on a raspberry pi etc.? (Not saying you have to optimize this one, but I would.)

8. Re "jumping somewhat unnecessary through hoops to satisfy the Rust compiler": Yeah, it's a bit more verbose but I'd always prefer to not have `unsafe`, especially if the safe way is not that painful :)

**Comment by dheijl** [comment]
OK thanks for your comments! 
Re 6. "... Vec size...":  where do I declare the Vec then? once_cell again?

**Comment by Boscop** [comment]
You would declare one before each of the 3 closures passed to `build_input_stream`
https://github.com/dheijl/swyh-rs/blob/a3f524862f069def61a7acf244cf817638032c75/src/main.rs#L823
and let the closure capture it. Instead of
```rust
            move |data, _: &_| wave_reader::<f32>(data), 
```
you'd pass this
```rust
            {
                let mut i16_samples = vec![];
                move |data, _: &_| wave_reader::<f32>(data, &mut i16_samples)
            },
```
and then
```rust
fn wave_reader<T>(samples: &[T], i16_samples: &mut Vec<i16>)
where
    T: cpal::Sample,
{
    static ONETIME_SW: OnceCell<()> = OnceCell::new();
    ONETIME_SW.get_or_init(|| {
        log("The wave_reader is receiving samples".to_string());
    });

    i16_samples.clear();
    i16_samples.extend(samples.iter().map(|x| x.to_i16()));
    let clients = CLIENTS.lock().unwrap();
    for (_, v) in clients.iter() {
        v.write(&i16_samples);
    }
}
```

Every closure is basically a struct that contains the fields corresponding to the variables it captures.
By moving `i16_samples` into the closure, it becomes an owned field, and then you can mutate it.
So each of the 3 closures has its own instance.

**Comment by Boscop** [comment]
Actually, since only one of the match arms gets executed, it's enough to define one Vec.
So you would declare it before the `match` statement, before this line:
https://github.com/dheijl/swyh-rs/blob/a3f524862f069def61a7acf244cf817638032c75/src/main.rs#L820

**Comment by dheijl** [comment]
Yes, that's the only solution that seemed feasible, glad that I was correct. Thanks!

**Comment by dheijl** [comment]
I think I have implemented all your suggestions now (I switched the Mutex to a parking_lot Mutex after reading up on it).

Thanks again!

**Comment by Boscop** [comment]
@dheijl Btw, since cpal recently changed their API I was trying to see which cpal version swyh uses, and I noticed that you use `"*"` for most deps, which is bad practice because it means when someone else checks out this repo who doesn't have your Cargo.lock file, they will get the most recent version of the deps and those newer versions might have breaking changes, so the build will fail.
So I recommend going through the deps and writing the actual version that is used (by looking in Cargo.lock).

2. It's recommended to add the Cargo.lock to the repo for executable crates, so that others who check out the repo will get the exact same dep versions, not just semver-compatible ones:
https://github.com/dheijl/swyh-rs/blob/2e6924c9108cb1f7d78d9245ba39ef06a6ad77e1/.gitignore#L6

**Comment by Boscop** [comment]
Btw, you don't need `extern crate` with edition 2018, especially not in files that aren't main.rs/lib.rs:
https://github.com/dheijl/swyh-rs/blob/2e6924c9108cb1f7d78d9245ba39ef06a6ad77e1/src/utils/audiodevices.rs#L1

**Comment by dheijl** [comment]
Thanks, I'll remove the extern crate line (remnant from cpal example) and add the cargo.lock.

I prefer to use the latest versions of everything (except if there are dependendency problems between the crates I use). I don't mind if an api change breaks something in my code, it's usually for the best.

**Comment by Boscop** [comment]
> I prefer to use the latest versions of everything

Don't we all :)
> I don't mind if an api change breaks something in my code, it's usually for the best.

I mean, if someone else checks out this repo and just wants to build it successfully, it's better UX if the build succeeds ;)
Btw, I use [`cargo upgrades`](https://crates.io/crates/cargo-upgrades) to check which deps have higher versions. And apart from `cargo update` (which only updates to semver-compatible versions), I also use [`cargo upgrade`](https://crates.io/crates/cargo-edit#cargo-upgrade) to upgrade deps to the latest available version (when I have time to deal with build breakage).

Summary:
- `cargo update` only modifies your `Cargo.lock` file
- `cargo upgrades` only shows available newer versions
- `cargo upgrade` actually modifies your `Cargo.toml` as when you manually upgrade a dependency

**Comment by dheijl** [comment]
Thanks for this advice, it saves me hours of reading documentation and seems to answer all the questions I could have on the subject!

**Comment by Boscop** [comment]
I'm glad to help :)
Yeah, it took me a while to discover all these utilities..
Another one I regularly use for bulk search/replace is https://crates.io/crates/amber

---

Btw, I just implemented the loopback capture like swyh in my visualizer. 
So far only tried running it on linux but I only get silence in, even if I'm playing music in chromium (on Manjaro).
(I'm not very familiar with Linux audio, mostly using Windows for audio.)
This is the default device apparently:
```
[INFO ] Capturing audio from: default
[INFO ] Default audio cfg: SupportedStreamConfig { channels: 2, sample_rate: SampleRate(44100), buffer_size: Range { min: 3, max: 4194304 }, sample_format: F32 }
```
This is the code:
```rust
pub fn capture_live_audio_input() -> Result<(BidirAudioChannel, AudioFormat)> {
	let (audio_to_vis, vis_to_audio) = BidirAudioChannel::duplex_unbounded();
	let output_device = unwropt!(get_default_audio_output_device(), "No default audio device");
	debug!("Try capturing system audio");
	capture_output_audio(&output_device, audio_to_vis).context("Could not capture audio").map(
		|(stream, audio_format)| {
			stream.play().unwrap();
			(vis_to_audio, audio_format)
		},
	)
}

fn capture_output_audio(
	output_device: &Device,
	audio_to_vis: BidirAudioChannel,
) -> Result<(Stream, AudioFormat)> {
	info!(
		"Capturing audio from: {}",
		output_device.name().context("Could not get default audio device name")?
	);

	let audio_cfg = output_device.default_output_config().context("No default output config found")?;
	let audio_format =
		AudioFormat { sample_rate: audio_cfg.sample_rate().0, channel_count: audio_cfg.channels() as usize };
	info!("Default audio cfg: {:?}", audio_cfg);

	fn build_stream<T: Sample>(
		output_device: &Device,
		audio_cfg: SupportedStreamConfig,
		audio_format: AudioFormat,
		audio_to_vis: BidirAudioChannel,
	) -> Result<(Stream, AudioFormat)> {
		fn wave_reader<T: Sample>(
			samples: &[T],
			f32_samples: &mut AudioFrame,
			audio_to_vis: &BidirAudioChannel,
		) {
			f32_samples.clear();
			f32_samples.extend(samples.iter().map(|x| x.to_f32()));
			let _ = audio_to_vis.tx.send(f32_samples.clone());
		}

		let mut f32_samples = Vec::with_capacity(1024);
		match output_device.build_input_stream(
			&audio_cfg.config(),
			move |data, _: &_| wave_reader::<T>(data, &mut f32_samples, &audio_to_vis),
			|e| error!("Error on audio input stream: {}", e),
		) {
			Ok(stream) => Ok((stream, audio_format)),
			Err(e) => {
				bail!("Error capturing {:?} audio stream: {}", T::FORMAT, e);
			}
		}
	}

	match audio_cfg.sample_format() {
		SampleFormat::F32 => build_stream::<f32>(output_device, audio_cfg, audio_format, audio_to_vis),
		SampleFormat::I16 => build_stream::<i16>(output_device, audio_cfg, audio_format, audio_to_vis),
		SampleFormat::U16 => build_stream::<u16>(output_device, audio_cfg, audio_format, audio_to_vis),
	}
}
```
Any idea why I'm only getting silence? :)

**Comment by dheijl** [comment]
The only similar problem I ever had was because a) I forgot to start to play the stream and b) the play statement went out of scope before the capture had even started.
Also, your code is rather dense for my old "Rust newbie" head... I had no idea that inner methods were supported.

**Comment by Boscop** [comment]
Ah, you mean the `stream` variable needs to be dropped after `stream.play()` is called?
Or even longer?

Btw, `get_output_audio_devices` prints this:
```
connect(2) call to /dev/shm/jack-1000/default/jack_0 failed (err=No such file or directory)
attempt to connect to server failed
connect(2) call to /dev/shm/jack-1000/default/jack_0 failed (err=No such file or directory)
attempt to connect to server failed
ALSA lib pcm_oss.c:377:(_snd_pcm_oss_open) Unknown field port
ALSA lib pcm_oss.c:377:(_snd_pcm_oss_open) Unknown field port
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_usb_stream.c:508:(_snd_pcm_usb_stream_open) Unknown field hint
ALSA lib pcm_usb_stream.c:508:(_snd_pcm_usb_stream_open) Unknown field hint
ALSA lib pcm_usb_stream.c:508:(_snd_pcm_usb_stream_open) Unknown field hint
ALSA lib pcm_usb_stream.c:508:(_snd_pcm_usb_stream_open) Unknown field hint
ALSA lib pcm_dmix.c:1090:(snd_pcm_dmix_open) unable to open slave
ALSA lib pcm_dmix.c:1090:(snd_pcm_dmix_open) unable to open slave
ALSA lib pcm_usb_stream.c:508:(_snd_pcm_usb_stream_open) Unknown field hint
ALSA lib pcm_usb_stream.c:508:(_snd_pcm_usb_stream_open) Unknown field hint
```
And these are the ones with a default config (returned by the function):
```
[
    "samplerate",
    "speexrate",
    "pipewire",
    "pulse",
    "speex",
    "upmix",
    "vdownmix",
    "default",
    "sysdefault:CARD=PCH",
    "front:CARD=PCH,DEV=0",
    "surround40:CARD=PCH,DEV=0",
    "surround51:CARD=PCH,DEV=0",
    "surround71:CARD=PCH,DEV=0",
    "iec958:CARD=PCH,DEV=0",
    "hdmi:CARD=NVidia,DEV=0",
    "hdmi:CARD=NVidia,DEV=1",
    "hdmi:CARD=NVidia,DEV=2",
    "hdmi:CARD=NVidia,DEV=3",
    "hdmi:CARD=NVidia,DEV=4",
    "hdmi:CARD=NVidia,DEV=5",
    "hdmi:CARD=NVidia,DEV=6",
]

```

**Comment by dheijl** [comment]
I meant the stream went out of scope, aborting the capture. 
I did not investigate Linux audio much, just made sure it compiled and worked with the default device.

**Comment by Boscop** [comment]
I added some `println`s and it seems that `build_stream` gets called and succeeds but `wave_reader` never gets called. Hm..

**Comment by Boscop** [comment]
Ah, yes, you're right, when I sleep the thread for 5s, `wave_reader` gets called for those 5s and then not anymore. So yes, the problem is that I let `stream` go out of scope. Thanks :)

**Comment by Boscop** [comment]
@dheijl Hm, for some reason now my `wave_reader` doesn't get called, even though `build_stream` succeeds.
I'd really appreciate if you could help me figure this out: https://github.com/RustAudio/cpal/issues/516

**Comment by dheijl** [comment]
A couple of thoughts:

WasApi is COM based, and cpal COM initialization has to be done before anything else to make sure that the correct threading model is being used, there's an open issue about that.

Also, I think that you have to call (some?) WasApi functions from the main (UI) thread, but not sure about that.

Other than that I would have to test your code to see what's going on.
